### PR TITLE
Ocp4 workload example - changes for class

### LINKED ISSUE #1869: Add Quota example to ocp-workload-example
**Is your feature request related to a problem? Please describe.**
Typical user will need to add quotas, too.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #1870: Add more robust example for ocp-workload-example
**Is your feature request related to a problem? Please describe.**
Deploy a pod and use user.info, for example.  Maybe something more complex, like an operator.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by newgoliath**
##### SUMMARY
Prep for first half of aD/V class.

* Leave ocp4_workload_example as originally.  Just emits a var.
* Create ocp4_workload_example_shared_cluster
** Deploys a sample app via yaml manifests as `ocp-username`

TODO:

* update for indicating OpenTLC credentials to shared bastion host to run playbook

Fixes #1869 and Fixes #1870 

##### ISSUE TYPE
- New role Pull Request

##### COMPONENT NAME
ocp4_workload_example
ocp4_workload_example_shared_cluster

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc [hunk 1]
```diff
@@ -22,25 +22,23 @@
 *** This role doesn't do anything here
 *** Debug task will print out: `remove_workload Tasks completed successfully.`
 
-== Review the defaults variable file
+== The defaults variable file
 
 * This file link:./defaults/main.yml[./defaults/main.yml] contains all the variables you need to define to control the deployment of your workload.
 * The variable *ocp_username* is mandatory to assign the workload to the correct OpenShift user.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:28]: Want to add the same comment as in the other file that this will be overridden by CloudForms?

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc [hunk 2]
```diff
@@ -22,25 +22,23 @@
...
 * A variable *silent=True* can be passed to suppress debug messages.
 * You can modify any of these default values by adding `-e "variable_name=variable_value"` to the command line
-* Add a dictionary for workload parameters.
+* Add long-name scoped workload parameters. Example: `ocp4_workload_example_machineconfigpool_name: worker`
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:31]: Thanks for fixing this!

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc [hunk 3]
```diff
@@ -22,25 +22,23 @@
...
 === Overview
 
-A `dictionary` approach should be used in order to neatly encapsulate parameters to roles.
-
-The role itself defines a dictionary `<role name>_defaults` in the file link:./defaults/main.yml[./defaults/main.yml]. Because this example role is `ocp-workload-example` the dictionary is called `ocp_workload_example_defaults`. Note that `_` is used instead of `-` in the dictionary name to create a valid Ansible dictionary name.
-
-When deploying a workload you can specify another dictionary with the specific parameters for this deployment. This dictionary should be called `<role name>_vars`. In our example this dictionary would be called `ocp_workload_example_vars`.
-
-
+Please use long-name scope parameters for your workload.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:37]: I'd rephrase to "You *must* use long-name scope parameters for your workload to avoid variable clashing"

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc [hunk 4]
```diff
@@ -22,25 +22,23 @@
...
-The role itself defines a dictionary `<role name>_defaults` in the file link:./defaults/main.yml[./defaults/main.yml]. Because this example role is `ocp-workload-example` the dictionary is called `ocp_workload_example_defaults`. Note that `_` is used instead of `-` in the dictionary name to create a valid Ansible dictionary name.
-
-When deploying a workload you can specify another dictionary with the specific parameters for this deployment. This dictionary should be called `<role name>_vars`. In our example this dictionary would be called `ocp_workload_example_vars`.
-
-
+Please use long-name scope parameters for your workload.  
+Ansible lacks robust variable scoping.
+For parameters passed into your workload role, prepend the role name.
+For parameters shared across workloads, have a catch-all name that applies to your goal.
+For example, parameters named `ocp4_workload_example_*` would apply to workloads `ocp4-workload-example` itself *and* `ocp4-workload-example-minio`.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:41]: Hm. I find that confusing.

ocp4_workload_example_* is for the ocp4_workload_example workload
ocp4_workload_example_minio_* is for a ocp4_workload_example_minio workload.

I think we should be *really* prescriptive here. No ambiguities

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc [hunk 5]
```diff
@@ -167,7 +128,7 @@ platform: rhpds
 cloud_provider: none
 
 # --- Config
-env_type: ocp-workloads
+env_type: ocp4-cluster
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:131]: Why this change? You use the config `ocp-workloads` to deploy workloads onto an existing cluster.

File: ansible/roles_ocp_workloads/ocp4_workload_example/tasks/remove_workload.yml [hunk 6]
```diff
@@ -2,8 +2,6 @@
 # Implement your workload removal tasks here
 # ------------------------------------------
 
-
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example/tasks/remove_workload.yml:5]: There's a reason I left the empty lines..... I wanted to make sure it was obvious where to add "things".

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc [hunk 7]
```diff
@@ -0,0 +1,253 @@
+= ocp4_workload_example_shared_cluster - Example Workload Role deployed to Shared Clusters
+
+== Preparation to Use This Role
+
+=== Get Access to Shared Cluster and Create Quotas
+
+This example role assumes you'll be using the Shared Cluster provided by GPTE.
+There should be a catalog item in CloudForms to give you access and get stated developing.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc:8]: Typo: stated -> started

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+
+== Description
+
+This role teaches the use of agnosticD ocp4_workload roles on pre-existing OpenShift clusters where admin access to the cluster is not available.
+The user must have "self-provisioner" roles and quotas already setup by another process, usually labs.opentlc.com or rhpds.redhat.com.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc:13]: That's true for developing. But AgnosticV will run these roles as "system:admin"

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+*** Debug task will print out: `post_workload Tasks completed successfully.`
+
+** Playbook: link:./tasks/remove_workload.yml[remove_workload.yml] - Used to delete the workload
+*** This role deletes the OpenShift projected created for the user.
+*** Debug task will print out: `remove_workload Tasks completed successfully.`
+
+=== The defaults variable file
+
+* This file link:./defaults/main.yml[./defaults/main.yml] contains all the variables you need to define to control the deployment of your workload.
+* The variable *ocp_username* is mandatory to assign the workload to the correct OpenShift user.
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc:48]: Again maybe add the comment that it will be overwritten by CF.

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+* A variable *silent=True* can be passed to suppress debug messages.
+* You can override any of these default values by adding `-e "variable_name=variable_value"` to the command line or through your deployer.
+* Note that this role users long-name scoped workload parameters. Example: `ocp4_workload_example_shared_cluster_project_name: my-project-name`, not just `project_name`
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc:51]: Typo: users -> uses

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+For parameters shared across workloads, have a catch-all name that applies to your goal.
+For example, parameters named `ocp4_workload_example_shared_cluster_*` would apply to workloads `ocp4-workload-example_shared_cluster` itself *and* `ocp4_workload_example_shared_cluster_minio`.
+
+=== Provide info to users
+
+In your playbooks, emit as part of a `debug` play the `user.info` variables to provide important info back to users via email.
+
+.For example:
+----
+- name: output workshop username distribution URL
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/readme.adoc:67]: Check with Johnathan. I think we want to use the user.info module for this...

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/post_workload.yml [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+# For RHPDS deployment (onto a shared cluster) set
+# workload_shared_deployment to True
+# (in the deploy script or AgnosticV configuration)
+- name: post_workload tasks complete
+  debug:
+    msg: "Post-Software checks completed successfully"
+  when:
+  - not silent|bool
+  - workload_shared_deployment|default(False)
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/post_workload.yml:29]: I think you want a new-line at the end.

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/workload.yml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    kind: Deployment
+    namespace: "{{ ocp4_workload_example_shared_cluster_project_name }}"
+    name: "hello-openshift"
+  register: r_hello_openshift_deployment
+  retries: 30
+  delay: 10
+  until:
+    - r_hello_openshift_deployment.resources | length | int > 0
+    - r_hello_openshift_deployment.resources[0].status.availableReplicas is defined
+    - r_hello_openshift_deployment.resources[0].status.availableReplicas | int == r_hello_openshift_deployment.resources[0].spec.replicas | int
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/workload.yml:34]: I think you want "readyReplicas"

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/workload.yml [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+- name: get route by querying the OpenShift API
+  k8s_info:
+    api_version: route.openshift.io/v1
+    kind: Route
+    name: "hello-openshift"
+    namespace: "{{ ocp4_workload_example_shared_cluster_project_name }}"
+  register: r_app_route
+
+- name: notify user of route with "user.info " in logs.  CloudForms will see this and send email to users
+  debug:
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/tasks/workload.yml:45]: Again. User.info module.

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/templates/namespace.j2 [hunk 15]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    openshift.io/description: ''
```
Comment (wkulhanek) [inline ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/templates/namespace.j2:5]: If you do all the openshift specific stuff why not create a project instead?

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc (hunk 4 above)
Comment (newgoliath) [reply ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:41]: Yes, I shouldn't yet address the situation where a deployment uses many workloads that share variables.

File: ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc (hunk 5 above)
Comment (newgoliath) [reply ansible/roles_ocp_workloads/ocp4_workload_example/readme.adoc:131]: Oops.. duh.  Reverting.

File: ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/templates/namespace.j2 (hunk 15 above)
Comment (newgoliath) [reply ansible/roles_ocp_workloads/ocp4_workload_example_shared_cluster/templates/namespace.j2:5]: /shrug  So many options.  Better to just create a project, sure.

**Comment by newgoliath** [comment]
@wkulhanek All addressed.  Please review again.
